packages/commerce-sdk-react/src/hooks/ShopperProducts/query.ts

```diff
@@ -7,21 +7,37 @@
 import {ApiClients, Argument, DataType, QueryResponse} from '../types'
 import {useAsync} from '../useAsync'
 import useCommerceApi from '../useCommerceApi'
+import {getDependencySource} from '../../utils/util'
 
 type Client = ApiClients['shopperProducts']
-
 /**
  * A hook for `ShopperProducts#getProducts`.
  * Allows access to multiple products by a single request. Only products that are online and assigned to a site catalog are returned. The maximum number of productIDs that can be requested are 24. Along with product details, the availability, images, price, promotions, and variations for the valid products will be included, as appropriate.
  * @see {@link https://developer.salesforce.com/docs/commerce/commerce-api/references/shopper-products?meta=getProducts} for more information about the API endpoint.
  * @see {@link https://salesforcecommercecloud.github.io/commerce-sdk-isomorphic/classes/shopperproducts.shopperproducts-1.html#getproducts} for more information on the parameters and returned data type.
  * @returns An object describing the state of the request.
  */
-export const useProducts = (
-    arg: Argument<Client['getProducts']>
-): QueryResponse<DataType<Client['getProducts']>> => {
+function useProducts(
+    parameters: NonNullable<Argument<Client['getProducts']>>['parameters'],
+    deps?: unknown[],
+    headers?: NonNullable<Argument<Client['getProducts']>>['headers'],
+    rawResponse?: false
+): QueryResponse<DataType<Client['getProducts']>>
+function useProducts(
+    parameters: NonNullable<Argument<Client['getProducts']>>['parameters'],
+    deps?: unknown[],
+    headers?: NonNullable<Argument<Client['getProducts']>>['headers'],
+    rawResponse?: true
+): QueryResponse<Response>
```

Comment:
I think having 4 arguments is not a good experience. For example: what if I want to pass in `true` for the `rawResponse`, and I don't have anything for the earlier arguments? I'd still need to pass in all 4 arguments. 

IMO it's better to have all of the optional arguments to be in one object.

Reply:
Since I think of `useProducts` as the _highest-level_ hook that we have, how about having it be very simple for developers to call? Like accepting only `parameters`, and that's it. The rest of the arguments we'll take care of them, so developers do not have to.

If they want more control, then they are free to use the lower-level hooks like `useAsync` and `useCommerceApi`. And set the headers, rawResponse, etc.

Reply:
By design, only the parameters are required, and the rest will be taken care of if they are undefined. The users do not really have to pass the optional parameters. Although we expose high-level, and low-level hooks, I think we would want the users to use our high-levels hooks more unless they really can't. If we only provide one simple use case for our high-level hooks, it is likely that the users might just rely on the low-level more, which could probably destroy the purpose of writing these hooks imo

I'd agree with you that using the optional arguments could be a better interface.

Reply:
@vmarta We've decided that we will use react-query as an underlying implementation. The interface would change as such. The `queryOptions` are the options we would pass to react-query's useQuery. Only the first argument is required for all the hooks in this PR
```
function useProduct(
    parameters: UseProductsParameters,
    options?: {
        headers: UseProductsHeaders
        rawResponse: false
        queryOptions: UseQueryOptions<DataType<Client['getProducts']> | Response, Error>
    }
): UseQueryResult<DataType<Client['getProducts']>, Error>
```

See the change to `useAsync` [here](https://github.com/SalesforceCommerceCloud/pwa-kit/blob/develop/packages/commerce-sdk-react/src/hooks/useAsync.ts#L11)

Reply:
Here we are discussing the interface of the hooks, and it's pretty important to get it right... let me chime in my thoughts...

I think we agree the first argument `parameters` is going to be an object which will be used to create cache key for the query, which means, if you have a parameter that the response depends on, you should add it to the first argument `parameters`.

For `headers` and `rawResponse`: do we expect the response to be different when headers/rawResponse value change? I think the answer is yes. So we need these two to be a part of the cache key.


```
function useProduct(
    parameters: {
        headers?: UseProductsHeaders
        rawResponse?: false
        ...REST_OF_UseProductsParameters
    }
    options?: UseQueryOptions<DataType<Client['getProducts']> | Response, Error>
): UseQueryResult<DataType<Client['getProducts']>, Error>
```

This also clearly separates the arguments to have distinct responsibility in terms of how we connect the libraries, this makes it easy to document.

1st argument `parameters` -> commerce-sdk-isomorphic
2nd argument `options` -> react-query